### ISSUE TITLE
Add `cargo install --update-all` as a simple way to update all locally installed tools with `cargo install`

### ISSUE THREAD

**Issue by eddyp** [rust-lang/cargo#9527]
**Describe the problem you are trying to solve**

As suggested in the second comment of #2082, it would be really useful to be able to easily upgrade all binary tools installed via `cargo install` in a system, now that #2082 was fixed for a single package (proposal in #6667 and implementation from #6798 stabilization via #7560).

**Describe the solution you'd like**

We currently have `cargo install --list` list information about the tools, but cargo appears to be lacking a way to upgrade them en-masse like this:

For instance, on a system I see there are some tools I have, but it's not clear if they are up to date:

```
C:\>cargo install --list
cargo-add v0.2.0:
    cargo-add.exe
cargo-asm v0.1.16:
    cargo-asm.exe
    cargo-llvm-ir.exe
cargo-audit v0.11.2:
    cargo-audit.exe
cargo-binutils v0.1.6:
    cargo-nm.exe
    cargo-objcopy.exe
    cargo-objdump.exe
    cargo-profdata.exe
    cargo-readobj.exe
    cargo-size.exe
    cargo-strip.exe
cargo-bloat v0.9.3:
    cargo-bloat.exe
cargo-deny v0.6.6:
    cargo-deny.exe
....
```

And the way I was thinking it would have worked was something like this (NOT AN ACTUAL LISTING):

```
C:\>cargo install --update-all
cargo-add v0.2.0: up to date
cargo-asm v0.1.16: up to date
cargo-audit v0.11.2: outdated, installing v0.14.1
cargo-binutils v0.1.6: outdated, instaling v0.3.3
cargo-bloat v0.9.3: outdated, installing v0.10.0
cargo-deny v0.6.6: outdated, installing v0.9.1
...
```

The `--update-all` option can have a "pretend/dry-run" suboption that coud just list what would be done, or whatever that flag name might be named in other places in cargo (suggestions for flag name, shamelessly copied from apt: -s, --simulate, --just-print, --dry-run, --recon, --no-act).

**Notes**

Not sure what should happen WRT semantic versioning, but my gut feeling under-0 versions could be updated by default, but the 1.0+ versions might make sense to have the behavior follow SemVer and not upgrade, so not sure which is a same default?

**Comment by weihanglo** [comment]
There exists a [`cargo-update` crate](https://crates.io/crates/cargo-update). Does it fit this use case?

**Comment by weihanglo** [comment]
For the SemVer compatibility cargo/crates.io repecting to, [The Cargo Book gets a nice chapter](https://doc.rust-lang.org/cargo/reference/semver.html#change-categories), and it seems that under `0.y.x` all `y` changes is considered as major release.

> This guide uses the terms "major" and "minor" assuming this relates to a "1.0.0" release or later. Initial development releases starting with "0.y.z" can treat changes in "y" as a major release, and "z" as a minor release. "0.0.z" releases are always major changes. This is because Cargo uses the convention that only changes in the left-most non-zero component are considered incompatible.

**Comment by eddyp** [comment]
> 
> 
> There exists a [`cargo-update` crate](https://crates.io/crates/cargo-update). Does it fit this use case?

Looking at the crate page look like the functionality is the desired one:

```cmd
C:\>cargo install-update -l
    Updating registry 'https://github.com/rust-lang/crates.io-index'

Package           Installed  Latest   Needs update
cargo-audit       v0.11.2    v0.14.1  Yes
cargo-binutils    v0.1.6     v0.3.3   Yes
cargo-inspect     v0.7.0     v0.10.3  Yes
cargo-llvm-lines  v0.4.6     v0.4.11  Yes
cargo-outdated    v0.9.9     v0.9.15  Yes
cargo-pants       v0.1.1     v0.1.25  Yes
...
cargo-deps        v1.4.1     v1.4.1   No
cargo-update      v7.0.1     v7.0.1   No
ripgrep           v12.1.1    v12.1.1  No
```

With its `cargo install-update-config` command seems to also handle the configuration of packages part that was mentioned in https://github.com/rust-lang/cargo/issues/2082#issuecomment-851346910 :

```
C:\>cargo install-update -h
cargo-install-update.exe-install-update 7.0.1
наб <nabijaczleweli@nabijaczleweli.xyz>
Yann Simon <yann.simon.fr@gmail.com>
ven <vendethiel@hotmail.fr>
Cat Plus Plus <piotrlegnica@piotrl.pl>
Liigo <liigo@qq.com>
azyobuzin <azyobuzin@users.sourceforge.jp>
Tatsuyuki Ishi <ishitatsuyuki@gmail.com>
Tom Prince <tom.prince@twistedmatrix.com>
Mateusz Mikuła <mati865@gmail.com>
sinkuu <sinkuupump@gmail.com>
Alex Burka <aburka@seas.upenn.edu>
Matthias Krüger <matthias.krueger@famsik.de>
Daniel Holbert <dholbert@cs.stanford.edu>
Jonas Bushart <jonas@bushart.org>
Harrison Metzger <harrisonmetz@gmail.com>
Benjamin Bannier <bbannier@gmail.com>
Dimitris Apostolou <dimitris.apostolou@icloud.com>
Corbin Uselton <corbinu@decimal.io>
QuarticCat <QuarticCat@protonmail.com>
A cargo subcommand for checking and applying updates to installed executables -- configuration

USAGE:
    cargo-install-update-config.exe install-update-config [FLAGS] [OPTIONS] <PACKAGE>

FLAGS:
    -a, --any-version               Allow any version
        --debug                     Compile the package in debug mode
        --enforce-lock              Require Cargo.lock to be up to date
    -h, --help                      Prints help information
        --install-prereleases       Install prerelease versions
        --no-enforce-lock           Don't enforce Cargo.lock
        --no-install-prereleases    Filter out prerelease versions
        --no-respect-binaries       Install all binaries
        --release                   Compile the package in release mode
        --respect-binaries          Only install already installed binaries

OPTIONS:
    -c, --cargo-dir <CARGO_DIR>                  The cargo home directory. Default: $CARGO_HOME or $HOME/.cargo
    -d, --default-features <DEFAULT_FEATURES>    Whether to allow default features
    -f, --feature <FEATURE>...                   Feature to enable
    -n, --no-feature <DISABLED_FEATURE>...       Feature to disable
    -t, --toolchain <TOOLCHAIN>                  Toolchain to use or empty for default
    -v, --version <VERSION_REQ>                  Require a cargo-compatible version range

ARGS:
    <PACKAGE>    Package to configure
```

and it seems `cargo install-update --all` does its job:

```cmd
C:\>cargo install-update --all
    Updating registry 'https://github.com/rust-lang/crates.io-index'

Package           Installed  Latest   Needs update
cargo-audit       v0.11.2    v0.14.1  Yes
cargo-binutils    v0.1.6     v0.3.3   Yes
cargo-inspect     v0.7.0     v0.10.3  Yes
cargo-llvm-lines  v0.4.6     v0.4.11  Yes
cargo-outdated    v0.9.9     v0.9.15  Yes
cargo-pants       v0.1.1     v0.1.25  Yes
cargo-udeps       v0.1.1     v0.1.21  Yes
git-absorb        v0.6.0     v0.6.6   Yes
git-bonsai        v0.1.0     v0.2.1   Yes
git-trim          v0.3.2     v0.4.2   Yes
hexyl             v0.5.1     v0.8.0   Yes
sd                v0.6.5     v0.7.6   Yes
svd2rust          v0.17.0    v0.19.0  Yes
tokei             v10.0.1    v12.1.2  Yes
topgrade          v5.8.1     v6.9.1   Yes
xargo             v0.3.22    v0.3.23  Yes
cargo-add         v0.2.0     v0.2.0   No
cargo-asm         v0.1.16    v0.1.16  No
cargo-bloat       v0.10.0    v0.10.0  No
cargo-deny        v0.9.1     v0.9.1   No
cargo-deps        v1.4.1     v1.4.1   No
cargo-update      v7.0.1     v7.0.1   No
ripgrep           v12.1.1    v12.1.1  No

Updating cargo-audit
  Downloaded cargo-audit v0.14.1
  Downloaded 1 crate (157.5 KB) in 1.16s
    Updating crates.io index
  Installing cargo-audit v0.14.1
  Downloaded addr2line v0.15.2
  Downloaded once_cell v1.7.2
  Downloaded stable_deref_trait v1.2.0
  Downloaded generational-arena v0.2.8
  Downloaded backtrace v0.3.60
  Downloaded object v0.25.2
  Downloaded tracing-core v0.1.18
  Downloaded unicode-normalization v0.1.19
  Downloaded zeroize v1.3.0
  Downloaded regex-automata v0.1.10
....

```



Just as a matter of taste, I kind of dislike the command names and the fact it prints the details of the authors upfront in `-h` invocations, but an integration into cargo proper could fix those.

The only reservation I have is the dependency on Cmake which might be a turnoff for some people, so it might make sense to integrate into cargo and drop the dependency:

```
Firstly, ensure you have CMake and the Required Libraries™.

Then proceed as usual:

    cargo install cargo-update
```

**Comment by Logarithmus** [comment]
@weihanglo what about integrating https://lib.rs/cargo-update into `cargo` itself?

**Comment by weihanglo** [comment]
I am not a member cargo team, so the followings are only my 2 cents. 

As Rust getting popular, the small cargo team cannot handle all these feature requests and bugs. They might not have extra time to take care more new commands. They need helps. If we do want to integrate [cargo-update](https://crates.io/crates/cargo-update) as a built-in command, there are things we can do:

- Learn how to organize and integrate a command into cargo, such as [cargo-tree] (integrated).
- Learn the possible obstacles we'll encounter during the integration, such as [cargo-edit] (ongoing).
- Learn what does the maintenance look like after integrating a new command, such as [cargo-vendor] (integrated).
- Discuss here or on [Zulip #t-cargo] about the design and idea of adding cargo-update. Also need to confirm whether or not an RFC is necessary.
- It will be pleasant if we can help maintain the new command for a reasonable period after the integration.

Looking forward to making cargo more friendly together!



[cargo-tree]: https://git.io/J1R5U
[cargo-edit]: https://git.io/JrgmL
[cargo-vendor]: https://github.com/rust-lang/cargo/issues?q=label%3ACommand-vendor+
[Zulip #t-cargo]: https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo

**Comment by martin-braun** [comment]
[This](https://stackoverflow.com/a/66049504/1540350) is a good work-around: 

```sh
cargo install $(cargo install --list | egrep '^[a-z0-9_-]+ v[0-9.]+:$' | cut -f1 -d' ')
```

**Comment by Rudxain** [comment]
> [This](https://stackoverflow.com/a/66049504/1540350)

I'm afraid the Cargo CLI API may not be stable enough to guarantee the regex doesn't break

**Comment by TaQuanMinhLong** [comment]
If you use [nushell](https://www.nushell.sh/) simply use this script

```nu
cargo install --list
    | parse "{package} v{version}:"
    | get package
    | each {|p| cargo install $p }
# Add --locked if you need locked
```
Edit: if you have [cargo-binstall](https://github.com/cargo-bins/cargo-binstall), this is much faster

```nu
def cargo_update [] {
    print "Updating cargo packages..."
    cargo install --list | parse "{package} v{version}:" | par-each { |i|
        echo $"Updating ($i.package)"
        ^cargo-binstall --no-confirm $i.package
    }
}
```

**Comment by jokeyrhyme** [comment]
`cargo install --list --json` would be amazing

**Comment by luccahuguet** [comment]
> If you use [nushell](https://www.nushell.sh/) simply use this script
> 
> cargo install --list
>     | parse "{package} v{version}:"
>     | get package
>     | each {|p| cargo install $p}

some crates only work with the locked option, so it might me better to go with 

```
 cargo install --list
     | parse "{package} v{version}:"
     | get package
     | each {|p| cargo install $p --locked}
```

**Comment by ardnew** [comment]
> [This](https://stackoverflow.com/a/66049504/1540350) is a work-around which doesn't require `cargo-update`:
> ```
> cargo install $(cargo install --list | egrep '^[a-z0-9_-]+ v[0-9.]+:$' | cut -f1 -d' ')
> ```

This should be sufficient for shells like bash/zsh:

```
cargo install $(cargo install --list|cut -sd' ' -f1)
```

All undesirable lines begin with whitespace, so `-f1` will yield blank lines for them, which are silently squashed via the command subst `$(...)`. 

If you're using bash with history support, you could be mean to future You with:

```
cargo install $(!#- --list|cut -sd' ' -f1)
```

**Comment by pcgeek86** [comment]
Here's a PowerShell command that will parse each crate installed with `cargo install`, and runs `cargo install` for it again.

```pwsh
cargo install --list | % { ($PSItem -match '^(?<name>\S+)\s') ? $matches.name : $null; } | % { cargo install $PSItem }
```

**Comment by rursprung** [comment]
> Here's a PowerShell command that will parse each crate installed with `cargo install`, and runs `cargo install` for it again.
> 
> cargo install --list | % { ($PSItem -match '^(?<name>\S+)\s') ? $matches.name : $null; } | % { cargo install $PSItem }

just FYI: this will not work in two cases:
* if the crate was installed with specific features (because you don't extract the features and then use them again on `cargo install`
* if something in the dependencies breaks (you can fix this by using `cargo install --locked`)

**Comment by Julienraptor01** [comment]
was really annoyed that there was no easy way to update packages so i made this bash alias which works quite well™
it's a bit horrendous and yeah [scary regex](https://jex.im/regulex/#!flags=&re=%5E(%5CS%2B)%5Cs%2Bv%5CS%2B(%5Cs%2B%5C((%5CS%2B%5C.git)%23%5CS%2B%5C))%3F%3A), but trust me, it's fairly simple and i'm sure there is no way it could ever break...
```bash
alias cargo-update='cargo install --list | awk '\''match($0, /^(\S+)\s+v\S+(\s+\((\S+\.git)#\S+\))?:/, m) {if (m[3]) {cmd="cargo install --git " m[3]} else {cmd="cargo install " m[1]}; system(cmd)}'\'''
```
<details>
<summary>Example</summary>

![Image](https://github.com/user-attachments/assets/df10afe6-fbb6-434a-bf0e-31dccf12db83)
</details>

**Comment by ST33LDI9ITAL** [comment]
@Julienraptor01 I believe that would break when package installed from git repo like fyn (https://github.com/duriantaco/fyn) from separate packages in workspace

**Comment by rsenna** [comment]
> was really annoyed that there was no easy way to update packages so i made this bash alias which works quite well™ it's a bit horrendous and yeah [scary regex](https://jex.im/regulex/#!flags=&re=%5E(%5CS%2B)%5Cs%2Bv%5CS%2B(%5Cs%2B%5C((%5CS%2B%5C.git)%23%5CS%2B%5C))%3F%3A), but trust me, it's fairly simple and i'm sure there is no way it could ever break...
> 
> alias cargo-update='cargo install --list | awk '\''match($0, /^(\S+)\s+v\S+(\s+\((\S+\.git)#\S+\))?:/, m) {if (m[3]) {cmd="cargo install --git " m[3]} else {cmd="cargo install " m[1]}; system(cmd)}'\'''

That's cool, but If you want an `awk` version, and you don't care so much about git repo crates, then it can get as simple as

```bash
cargo install --list | awk '/.* v.*/ { print $1 }'
```

So, for bash/zsh, we could do this:

```bash
alias cargo-list-packages="cargo install --list | awk '/.* v.*/ { print \$1 }'"
alias cargo-update='cargo install $(cargo-list-packages)'
alias cargo-bupdate='cargo binstall --no-confirm $(cargo-list-packages)'
```

For `fish`, using`function` is arguably less convoluted then `alias`:

```fish
function cargo-list-packages; cargo install --list | awk '/.* v.*/ { print $1 }'; end;
function cargo-update; cargo install (cargo-list-packages); end;
function cargo-bupdate; cargo binstall --no-confirm (cargo-list-packages); end;
```

Notes:
- Add `--locked` to taste.
- The `cargo-update` crate indeed seems like a better alternative. But, alas, I'm working on a constrained environment, where its compilation is so far broken.
- The `cargo-bupdate` version seems faster  _even without using explicit parallelisation_ - as seen in [@TaQuanMinhLong's cool nushell solution](https://github.com/rust-lang/cargo/issues/9527#issuecomment-2313951709).

**Comment by Julienraptor01** [comment]
\- @ST33LDI9ITAL
> [@Julienraptor01](https://github.com/Julienraptor01) I believe that would break when package installed from git repo like fyn ([duriantaco/fyn](https://github.com/duriantaco/fyn?rgh-link-date=2026-05-12T11%3A58%3A50.000Z)) from separate packages in workspace

fixed, just needed to add ` " " m[1]` after `m[3]`
and yet actually i revamped the snippet to handle literally every single possible cargo install

```bash
alias cargo-update='cargo install --list | awk '\''match($0, /^((\S+)\s+v\S+(\s+\((\S+?)(\?(branch|tag|rev)=(\S+?))?(#\S+)?\))?):/, m) { if (!m[4]) { cmd="cargo install " m[2] } else if (!m[8]) { print "\033[33mWarning: Skipping " m[1] "\033[0m" > "/dev/stderr"; cmd="" } else if (m[6]) { cmd="cargo install --git " m[4] " --" m[6] " " m[7] " " m[2] } else { cmd="cargo install --git " m[4] " " m[2] }; if (cmd) system(cmd) }'\'''
```

note : the lazy regex modifier is only available starting with awk 5.4.0 so if you're still on 5.3.2, use this
also very funni but gawk doesn't support \s inside of [^] so we need to use the posix [:space:]

```bash
alias cargo-update='cargo install --list | awk '\''match($0, /^((\S+)\s+v\S+(\s+\(([^[:space:]?#)]+)(\?(branch|tag|rev)=([^[:space:]#)]+))?(#[^[:space:])]+)?\))?):/, m) { if (!m[4]) { cmd="cargo install " m[2] } else if (!m[8]) { print "\033[33mWarning: Skipping " m[1] "\033[0m" > "/dev/stderr"; cmd="" } else if (m[6]) { cmd="cargo install --git " m[4] " --" m[6] " " m[7] " " m[2] } else { cmd="cargo install --git " m[4] " " m[2] }; if (cmd) system(cmd) }'\'''
```

@rsenna i do actually care about git crates

in an ideal world i wouldn't install git crates or even cargo as cli tools would be available in system package managers where they belong, but we are where we are... if at least i could have a convenient way to install and update those without needing to install even more stuff it'd be great, being forced to not only install cargo, but also cargo-update, just to have a tool and have it updated simply is quite silly, if cargo end up not just being a dependency manager, but also a package manager, it'd be great if it could manage

TL;DR please rust people, just make your manager have an update command built-in thanks <3

**Comment by jokeyrhyme** [comment]
that does bring to mind an alternative, which is to officially deprecate `cargo install`, to make it clear that `cargo` should not be used as a package manager 

i'd personally prefer cargo become a fully feature package manager instead, but explicitly avoiding this space is a path the project _could_ take
